File: TheWatch-iOS/Logging/Native/NativeLoggingAdapter.swift
import Foundation
import SwiftData
import Combine
import os.log

/// Native logging adapter — Tier 2 (On-Device + Offline).
///
/// - Persists every log entry to SwiftData (local SQLite)
/// - Echoes Warning+ to os_log for crash diagnostics
/// - Emits to a Combine publisher for real-time observation
/// - Works fully offline — no network dependency
/// - Paired with `NativeLogSyncAdapter` for Firestore sync
@MainActor
final class NativeLoggingAdapter: LoggingPort, @unchecked Sendable {

    private let modelContainer: ModelContainer
    private let subject = PassthroughSubject<LogEntry, Never>()
    private let osLog = Logger(subsystem: "com.thewatch.app", category: "NativeLog")

    var logPublisher: AnyPublisher<LogEntry, Never> {
        subject.eraseToAnyPublisher()
    }

    init(modelContainer: ModelContainer) {
        self.modelContainer = modelContainer
    }

    func write(_ entry: LogEntry) async {
        let model = LogEntryModel.fromDomain(entry)
        let context = ModelContext(modelContainer)
        context.insert(model)
        try? context.save()

        // Echo Warning+ to os_log
        if entry.level >= .warning {
            let msg = "[\(entry.sourceContext)] \(entry.renderedMessage)"
            switch entry.level {
            case .warning: osLog.warning("\(msg)")
            case .error:   osLog.error("\(msg)")
            case .fatal:   osLog.critical("\(msg)")
            default: break
            }
        }

        subject.send(entry)
    }

    func query(limit: Int = 100, minLevel: LogLevel = .verbose, sourceContext: String? = nil) async -> [LogEntry] {
        let context = ModelContext(modelContainer)
        var descriptor = FetchDescriptor<LogEntryModel>(
            predicate: #Predicate { $0.level >= minLevel.rawValue },
            sortBy: [SortDescriptor(\.timestamp, order: .reverse)]
        )
        descriptor.fetchLimit = limit

        do {
            var results = try context.fetch(descriptor)
            if let source = sourceContext {
                results = results.filter { $0.sourceContext == source }
            }
            return results.map { $0.toDomain() }
        } catch {
            osLog.error("[NativeLogging] Query failed: \(error.localizedDescription)")
            return []
        }
    }

    func queryByCorrelation(_ correlationId: String) async -> [LogEntry] {
        let context = ModelContext(modelContainer)
        let descriptor = FetchDescriptor<LogEntryModel>(
            predicate: #Predicate { $0.correlationId == correlationId },
            sortBy: [SortDescriptor(\.timestamp)]
        )

        do {
            return try context.fetch(descriptor).map { $0.toDomain() }
        } catch {
            osLog.error("[NativeLogging] Correlation query failed: \(error.localizedDescription)")
            return []
        }
    }

    func flush() async {
        let context = ModelContext(modelContainer)
        let total = (try? context.fetchCount(FetchDescriptor<LogEntryModel>())) ?? 0
        let unsyncedDescriptor = FetchDescriptor<LogEntryModel>(
            predicate: #Predicate { $0.synced == false }
        )
        let unsynced = (try? context.fetchCount(unsyncedDescriptor)) ?? 0
        osLog.debug("[NativeLogging] Flush — \(total) total, \(unsynced) unsynced")
    }

    func prune(olderThan date: Date) async {
        let context = ModelContext(modelContainer)
        do {
            let descriptor = FetchDescriptor<LogEntryModel>(
                predicate: #Predicate { $0.timestamp < date }
            )
            let old = try context.fetch(descriptor)
            for entry in old {
                context.delete(entry)
            }
            try context.save()
            osLog.debug("[NativeLogging] Pruned \(old.count) entries")
        } catch {
            osLog.error("[NativeLogging] Prune failed: \(error.localizedDescription)")
        }
    }
}

File: TheWatch-Android/app/src/main/java/com/thewatch/app/data/logging/native/NativeLoggingAdapter.kt
package com.thewatch.app.data.logging.native

import android.util.Log
import com.thewatch.app.data.logging.LogEntry
import com.thewatch.app.data.logging.LogLevel
import com.thewatch.app.data.logging.LoggingPort
import com.thewatch.app.data.logging.local.LogEntryDao
import com.thewatch.app.data.logging.local.LogEntryEntity
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.filter
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Native logging adapter — Tier 2 (On-Device + Offline).
 *
 * - Persists every log entry to Room (local SQLite)
 * - Echoes Warning+ to Logcat for crash diagnostics
 * - Emits to a SharedFlow for real-time observation
 * - Works fully offline — no network dependency
 * - Paired with [NativeLogSyncAdapter] for Firestore sync
 *
 * Room handles all threading via suspend functions on the IO dispatcher.
 */
@Singleton
class NativeLoggingAdapter @Inject constructor(
    private val dao: LogEntryDao
) : LoggingPort {

    private val liveFlow = MutableSharedFlow<LogEntry>(extraBufferCapacity = 256)

    companion object {
        private const val TAG = "TheWatch"
    }

    override suspend fun write(entry: LogEntry) {
        // Persist to Room
        dao.insert(LogEntryEntity.fromDomain(entry))

        // Echo Warning+ to Logcat for crash report correlation
        if (entry.level >= LogLevel.Warning) {
            val msg = "[${entry.sourceContext}] ${entry.renderedMessage()}"
            when (entry.level) {
                LogLevel.Warning -> Log.w(TAG, msg)
                LogLevel.Error   -> Log.e(TAG, msg)
                LogLevel.Fatal   -> Log.wtf(TAG, msg)
                else -> { /* unreachable */ }
            }
        }

        // Emit to observers
        liveFlow.emit(entry)
    }

    override suspend fun query(
        limit: Int,
        minLevel: LogLevel,
        sourceContext: String?
    ): List<LogEntry> {
        val entities = if (sourceContext != null) {
            dao.getRecentBySource(limit, minLevel.ordinal, sourceContext)
        } else {
            dao.getRecent(limit, minLevel.ordinal)
        }
        return entities.map { it.toDomain() }
    }

    override suspend fun queryByCorrelation(correlationId: String): List<LogEntry> {
        return dao.getByCorrelation(correlationId).map { it.toDomain() }
    }

    override fun observe(minLevel: LogLevel): Flow<LogEntry> {
        return liveFlow.filter { it.level >= minLevel }
    }

    override suspend fun flush() {
        // Room auto-persists, but log the buffer state for diagnostics
        val total = dao.count()
        val unsynced = dao.unsyncedCount()
        Log.d(TAG, "[NativeLogging] Flush — $total total, $unsynced unsynced")
    }

    override suspend fun prune(olderThanEpochMillis: Long) {
        val deleted = dao.deleteOlderThan(olderThanEpochMillis)
        Log.d(TAG, "[NativeLogging] Pruned $deleted entries older than $olderThanEpochMillis")
    }
}
